### PR TITLE
Moving Standard deviation

### PR DESCRIPTION
**Pull request by Bartoshko**
I am going to use this library in my Rust Rocket Server, as this is nice implementation of technical analysis algorithms. Well done! I have seen that it would be nice to have some more indicators, so I have added what I need the most, which is Moving Standard Deviation.  And if this project is open for contribution I would be happy to do some work, and propose what would be a 'nice to have'. Thx.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,7 @@
 /target/
 **/*.rs.bk
-Cargo.lock
+Cargo.loc
+
+# vim cached rescue files
+**.swp
+**.swo
```
Comment (greyblake) [inline .gitignore]: Can you please revert these changes?
You can configure Vim to keep swap files in a separate directory if you add something like this to your `vimrc`:

```
set dir=$HOME/.nvim/tmp
```

File: Cargo.lock [hunk 2]
```diff
@@ -0,0 +1,135 @@
+# This file is automatically @generated by Cargo.
```
Comment (greyblake) [inline Cargo.lock]: This file in gitignore and should not be committed

File: src/indicators/maximum.rs [hunk 3]
```diff
@@ -90,8 +90,8 @@ impl<'a, T: High> Next<&'a T> for Maximum {
 
 impl Reset for Maximum {
     fn reset(&mut self) {
-        for i in 0..self.n {
-            self.vec[i] = -INFINITY;
+        for elem in self.vec.iter_mut() {
+            *elem = -INFINITY;
```
Comment (greyblake) [inline src/indicators/maximum.rs]: What was the reason to changes this indicator?

File: src/indicators/minimum.rs [hunk 4]
```diff
@@ -90,8 +90,8 @@ impl<'a, T: Low> Next<&'a T> for Minimum {
 
 impl Reset for Minimum {
     fn reset(&mut self) {
-        for i in 0..self.n {
-            self.vec[i] = INFINITY;
+        for elem in self.vec.iter_mut() {
+            *elem = INFINITY;
```
Comment (greyblake) [inline src/indicators/minimum.rs]: What was the reason to changes this indicator?

File: src/indicators/moving_standard_deviation.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+impl Default for MovingStandardDeviation {
+    fn default() -> Self {
+        Self::new(9).unwrap()
+    }
+}
+
+impl fmt::Display for MovingStandardDeviation {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "SD({})", self.n)
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: MSD?

File: src/indicators/relative_strength_index.rs [hunk 6]
```diff
@@ -77,14 +77,20 @@ pub struct RelativeStrengthIndex {
...
+                    n: n,
+                    up_ema_indicator: Ema::new(n).unwrap(),
+                    down_ema_indicator: Ema::new(n).unwrap(),
+                    prev_val: 0.0,
+                    is_new: true,
+                };
+                Ok(rsi)
+            }
+        }
+
```
Comment (greyblake) [inline src/indicators/relative_strength_index.rs]: What was the reasons for the change? Could you please revert?

File: src/indicators/simple_moving_average.rs [hunk 7]
```diff
@@ -94,8 +94,8 @@ impl Reset for SimpleMovingAverage {
         self.index = 0;
         self.count = 0;
         self.sum = 0.0;
-        for i in 0..(self.n as usize) {
-            self.vec[i] = 0.0;
+        for elem in self.vec.iter_mut() {
+            *elem = 0_f64;
```
Comment (greyblake) [inline src/indicators/simple_moving_average.rs]: Could you please revert?

**Comment by greyblake** [comment]
@Bartoshko Thank you for your pull request. I left number of comments.
However I haven't yet deeply reviewed the implementation of the indicator itself.

Regarding the indicator Moving Standard Deviation, I understand it as the math concept, but never really have seen something like this in any of trading tools. E.g. in investing..com: https://www.investing.com/equities/twitter-inc-chart
Can you provide some examples of tools that use it?

From a first look you PR looks good, thanks for writing the proper doc and tests.

>  And if this project is open for contribution I would be happy to do some work, and propose what would be a 'nice to have'. Thx.

The project is open for contributions. However the goal of the project is to provide technical analysis indicators with a unified interfaces and good documentation.
In case if you really want to something add/implement I would suggest first to open an issue and initiate discussion to prevent useless effort if by some reason the change can not be accepted from the very beginning.

Thank you! =)

File: src/indicators/moving_standard_deviation.rs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+///
+/// * [Standard Deviation, Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation)
+///
+#[derive(Debug, Clone)]
+pub struct MovingStandardDeviation {
+    n: u32,
+    index: usize,
+    count: u32,
+    sum: f64,
+    vec: Vec<f64>,
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: `vec` could you please use another name (e.g. `values` or `prices`).
`vec` could be a little confusing because of the macro `vec!`.

File: src/indicators/moving_standard_deviation.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    
+    #[test]
+    fn test_next() {
+        let mut msd = MovingStandardDeviation::new(4).unwrap();
+        assert_eq!(msd.next(1.0), 1.7320508075688772);
+        assert_eq!(msd.next(2.0), 1.5811388300841898);
+        assert_eq!(msd.next(3.0), 1.4142135623730951);
+        assert_eq!(msd.next(4.0), 1.118033988749895);
+        assert_eq!(msd.next(10.0), 3.112474899497183);
+        assert_eq!(msd.next(20.0), 6.7592529172978875);
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: Are there some online tools that can prove correctness of this test data?

Could you please use [round()](https://github.com/greyblake/ta-rs/blob/master/src/test_helper.rs#L72) test helper, instead of writing very long floats? (see tests for some other indicators for examples)

File: src/indicators/moving_standard_deviation.rs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    #[test]
+    fn test_new() {
+        assert!(MovingStandardDeviation::new(0).is_err());
+        assert!(MovingStandardDeviation::new(1).is_ok());
+    }
+    
+    #[test]
+    fn test_next() {
+        let mut msd = MovingStandardDeviation::new(4).unwrap();
+        assert_eq!(msd.next(1.0), 1.7320508075688772);
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: I would expect that a SD for a data set, that contains only one item `1.0`, should be zero.

File: src/indicators/moving_standard_deviation.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        assert_eq!(msd.next(2.0), 1.5811388300841898);
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: According to [this online](http://www.alcula.com/calculators/statistics/standard-deviation/) tool SD of dataset `1, 2` must be `0.5`

image


Looks like the implementation is not correct.

File: src/indicators/moving_standard_deviation.rs [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        assert_eq!(msd.next(3.0), 1.4142135623730951);
+        assert_eq!(msd.next(4.0), 1.118033988749895);
+        assert_eq!(msd.next(10.0), 3.112474899497183);
+        assert_eq!(msd.next(20.0), 6.7592529172978875);
+    }
```
Comment (greyblake) [inline src/indicators/moving_standard_deviation.rs]: Can you please also add a test for a case when only input values are the same?
E.g.
```
5.4, 5.4, 5.4, 5.4
```

I would expect SD to be `0`

Comment (Bartoshko) [reply src/indicators/moving_standard_deviation.rs]: There is misunderstanding of this issue from my site. I have been thinking to use SD in Bollinger Bands, I see now that it isn't going to work for preferred usage of SD.

**Comment by Bartoshko** [comment]
I have read all your comments and I have to say I do agree with You. Therefore I don't see any value of going forward with what I have done. I just jumped in to this with a bit of rush. Thanks for Your time spent on QA ,I hope to build on this in the future. 
My point was to add some code that will help me with creating tool for Bollinger Bands analysis. 
I will rethink all and get back to you with an proper issue.
Thanks a lot =)

File: .gitignore (hunk 1 above)
Comment (Bartoshko) [reply .gitignore]: thanks, I will do.

File: src/indicators/moving_standard_deviation.rs (hunk 11 above)
Comment (Bartoshko) [reply src/indicators/moving_standard_deviation.rs]: This implementation was seeding population of n elements with `0_f64` instead of calculating only for inserted values. So this was going to properly function after n-th `next()` function call, I' am going to implement this correctly. Thanks for pointing this out.
